### PR TITLE
fix(restrict-hostnames): account for ingress or virtualservice with no path

### PR DESCRIPTION
**Pull request by frazs**
Addressing CN-1121.

Done:
- Fix existing test cases by adding a placeholder for `input.parameters.errorMsgAdditionalDetails`
- Add a test case for the example pathless ingress that escapes the policies
- Adjust path check so that the ingress no longer escapes
- Add a test case for a pathless virtualservice that escapes the policies

But:
- The path check fix that worked for the ingress makes other virtualservice test cases fail
  - It appears to consider the ingress and virtualservice as having an empty path *in addition* to any existing paths.
  - I confirmed this by adding an also now failing test case for an ingress that is limited to, and using, an allowed path

Done After:
- Confirm that the issue also impacts the hostname conflict violation (adjusted one of its test cases to have no path)
- Adjust the path checks so that all tests succeed. 
  - This was done with an `object.get` for the Ingress violation.
  - For the VirtualService and Hostpath Conflict violation, the only working way I've found thus far is to create a near-duplicate violation that checks that there are no paths and in that case validates against "" instead of passing the path variable.
- Copy finalized rego over to the template

### CODE REVIEW

**Comment by frazs** [comment]
Latest commits feature a working but very redundant solution. Now that there's a starting point passing all test cases, I (or anyone else interested!) can look into reducing the repetition.

**Comment by frazs** [comment]
Since my last comment: 
- Figured out a way to check for a pathless Ingress in the same violation using `object.get`. However this does not seem to support looking at multiple keys, as is needed when VirutalServices are involved.
- Confirmed that the issue impacts the Hostpath Conflict violation as well and produced a similar repetitive fix patterned after what worked for the VirtualService.

The present solution is working, but I am not sure if there is a more elegant way to go about it that I haven't found.

File: general/restrict-hostnames/rego/src.rego [hunk 1]
```diff
@@ -134,3 +153,36 @@ violation[{"msg": msg}] {
...
+# Pathless hostname conflict with other namespace Ingress(es) or VirtualService(s) and hostpath not allowed
+violation[{"msg": msg}] {
+	kind := input.review.kind.kind
+	re_match("^(Ingress|VirtualService)$", kind)
+	re_match("^(networking.k8s.io|networking.istio.io)$", input.review.kind.group)
+
+	hosts := {host | host := input.review.object.spec.rules[_].host} | {host | host := input.review.object.spec.hosts[_]}
+	host := hosts[_]
+	paths := ({path | path := input.review.object.spec.rules[_].http.paths[_].path} | ({path | path := input.review.object.spec.http[_].match[_].uri.exact} | {path | path := input.review.object.spec.http[_].match[_].uri.prefix})) | {path | path := input.review.object.spec.http[_].match[_].uri.regex}
+	count(paths) = 0
```
[the remark this answers is not in the document]
Comment (justbert) [reply general/restrict-hostnames/rego/src.rego:166]: Agreed here. Let's shit it out if it's working well.

File: general/restrict-hostnames/rego/src_test.rego [hunk 2]
```diff
@@ -1242,8 +1291,132 @@ test_allowed_vs_hostname_conflicts {
...
+	}}
+
+	exemptions := [""]
+
+	result := violation with input as vs_review with data.inventory.cluster.v1.Namespace as namespaces with input.parameters.exemptions as exemptions with input.parameters.errorMsgAdditionalDetails as "(Additional details placeholder)"
 
-	#Empty set means no violations
+	# Empty set means no violations
 	result == set()
 }
```
Comment (justbert) [inline general/restrict-hostnames/rego/src_test.rego:1422]: I've added this test case which should pass since it is `test.com` and `test.com/` is effectively the same for our rules. Dunno if we want to address this or just make a note and fix it in another ticket.

Comment (justbert) [reply general/restrict-hostnames/rego/src_test.rego:1422]: @frazs @MichaelPatsula

File: general/restrict-hostnames/rego/src.rego (hunk 1 above)
[the remark this answers is not in the document]
Comment (frazs) [reply general/restrict-hostnames/rego/src.rego]: Great catch Michael, updated to == 😄

File: general/restrict-hostnames/rego/src_test.rego (hunk 2 above)
Comment (frazs) [reply general/restrict-hostnames/rego/src_test.rego:1422]: Per the latest commit, modifying the new violations (and existing Ingress violation modified with `object.get`) so that when no path is found, the path argument is passed as `"/"` instead of `""` appears to satisfy this test case without messing up any of the other ones

Comment (frazs) [reply general/restrict-hostnames/rego/src_test.rego:1422]: Also confirmed that even with the change to the argument, this new test case is still happy if the allowed annotation uses `""` instead of `"/"` (which I think is a case `test_ingress_empty_path` already handles, so it probably doesn't warrant a new test case).

Comment (MichaelPatsula) [reply general/restrict-hostnames/rego/src_test.rego:1422]: For the no path edge case, currently there are the test cases:
- test_ingress_unallowed_host_no_path
- test_vs_unallowed_host_no_path 
- test_vs_allowed_no_path

Maybe we should also have a test case for test_ingress_allowed_no_path.

Comment (frazs) [reply general/restrict-hostnames/rego/src_test.rego]: In with the latest commit, if it looks good this can be resolved and the PR merged 😄

**Review by MichaelPatsula — APPROVED** [review]
LGTM!
